corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
         return value
     else:
         return None
+
+
+class UserChangeLogger(object):
+    """
+    User change logger to record
+        changes in user fields
+        messages for changes or
+        useful info text for user display like names
+    """
+    def __init__(self, domain, user, is_new_user, changed_by_user, changed_via):
+        self.domain = domain
+        self.user = user
+        self.is_new_user = is_new_user
+        self.changed_by_user = changed_by_user
+        self.changed_via = changed_via
+
+        self.original_user_doc = self.user.to_json()
+        self.fields_changed = {}
+        self.messages = []
+
+        self._save = False  # flag to check if log needs to be saved for updates
+
+    def add_changes(self, changes):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        for name, new_value in changes.items():
+            if self.original_user_doc[name] != new_value:
+                self.fields_changed[name] = new_value
+                self._save = True
+
+    def add_change_message(self, message):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        self.messages.append(message)
+        self._save = True
+
+    def add_info(self, info):
+        # useful info for display like names
+        self.messages.append(info)
+        self._save = True
+
+    def save(self):
```

Comment:
nit: how about changing this to `save_log` since this isn't a model itself.

Reply:
good point. This was named originally as save because it was handling user save and then log save if needed, but then because of multiple user.save calls outside of web importer, for web user uploads, it didn't make sense for user.save to be here, though the method name persisted.

Considering that state now, I am thinking this would be better as a context manager so that the logs are automatically saved on exit.
```
with CommCareUserImporter() as importer:
    # update first name
    # update role
    # etc etc
```
Renamed to `save_log` for now.

Reply:
https://github.com/dimagi/commcare-hq/pull/29935/commits/c7e526f7f23045a00126aa4456157fbf80ac45c9

Reply:
It doesn't look like this function name got changed, only the ones in the importer classes.